Handle filenames with slashes correctly

CraigFe opened the pull request:
This is a candidate solution for #160 and #115.

If any of the tests fail registration, Alcotest will search for any additional registration errors, print them to stderr, and then exit with code 1. Alcotest now also gives an error if a test name contains forward or backward slashes (as these names must be valid folder names -- see #115).

Original behaviour:
```
[ craigfe@carbon ] ~/t/alcotest
ᐅ dune exec test/main.exe
Error: "\239\129\173" is not a valid test label (it should be an ASCII string), skipping.
Testing alcotest.
This run has ID `CC560E17-DF73-485F-A026-EF15BAEBB9B1`.
[OK]                bar1                0   unix.
[OK]                foo1                0   lwt.
 ...                with/slash          0   name_with_slash/foo.main.exe: internal error, uncaught exception:
          Unix.Unix_error(Unix.ENOENT, "open", "/home/craigfe/t/alcotest/_build/_tests/CC560E17-DF73-485F-A026-EF15BAEBB9B1/with/slash.000.output")
```

New behaviour:
```
[ craigfe@carbon ] ~/t/alcotest
ᐅ dune exec test/main.exe
Error: "\239\129\173" is not a valid test label (it should be an ASCII string).
Error: "with/slash" is not a valid test label (contains an illegal character).

[ craigfe@carbon ] ~/t/alcotest
ᐅ echo $?
1
```

hannesm commented:
it is not immediately clear to me whether this is a complete solution for #115, i.e. have you thought about:
- test name being `.`
- test name being `..`
- two test names being case-insensitive equal (`foo`, `fOo`), and executing on a case-insensitive file system (IIRC macos still ship with such a thing by default)
- any other control character 0x00..0x20?

CraigFe commented:
It's a good concern to have; you're right that these cases are probably not being handled correctly. I had a look some sort of filename validation in [Filename](https://caml.inria.fr/pub/docs/manual-ocaml/libref/type_Filename.html), but couldn't find any. 

If possible, I'd like to avoid considering these edge cases explicitly in Alcotest. Do you have any suggestions?

hannesm commented:
Well, my main point was to figure out what is the problem space here. 
- A simple solution (breaking change) would be to not use test suite names as filenames, but instead numbers or hashes thereof.
- Another solution may be to have a whitelist of allowed characters?
- A third solution would be to url-percent-escape (%7F) "evil"/"bad" characters (or String.escape?).

I don't have sufficient time to figure out the design tradeoffs, which would involve to potentially break a lot of user intuition -- it may be worth to do some empirical evaluation of "bad/evil characters/strings used in test suite names of alcotest" first!?

hannesm commented:
(NB of course this PR is already an improvement of the current state, but I'm usually more interested in solutions that solve an issue once and for all -- excuse me for being pedantic ;)

samoht commented on src/alcotest.ml:
```diff
@@ -463,16 +463,25 @@ let list_tests t () =
-  in
-  Fmt.(pf stderr) "%a %s\n%!" red "Error:" err
+let contains_illegal_char s =
+  let disallowed = ['/'; '\\'] in
```
I think I'd prefer to have a whitelist, e.g. change `is_ascii` to actually check that the test name contains only "valid" characters (e.g. `a-z` + `A-Z`  + `0-9` + a few chars (maybe `.` and '-`' and `_`). (and not `..`)

samoht replied:
Actually I'm fine to disallow `.` as well. Can be fixed easily in most of my libs anyway :-)

CraigFe replied:
Agreed. This doesn't deal with the case-insensitive filesystem problem mentioned by @hannesm (which I haven't tested, but seems plausible), but we can worry about that separately.

CraigFe replied:
Will wait until #157 has been merged to make use of the `re` depencency.

CraigFe commented:
- Rebased
- Change to use a whitelist of characters instead
- Added a line in the README explaining the new limitation

Also:
- Changed the selective range syntax to `..` rather than `-` (on the suggestion of @cfcs [here](https://github.com/mirage/alcotest/issues/156#issuecomment-516049169)), to bring it in line with the README. Forgot to do this in #157... sorry!

samoht commented:
Thanks! I think we should also probably normalise filenames at one point, but I am not sure exactly how, and this could be done later.

samoht commented:
After a bit more thoughts, I am not sure breaking existing tests is a good idea -- and that people should be free to use whatever unicode characters as they want to name their tests.

So instead, we should make sure the filename are normalised properly, by replacing weird characters by something different (for instance `-`), something like:

```ocaml
let slugify s =
  let last_char_is_dash = ref false in
  let s = String.Ascii.lowercase s in
  let buf = Buffer.create (String.length s) in
  String.iter
    (function
      | ('a' .. 'z' | '0' .. '9') as c ->
        if !last_char_is_dash then (
          Buffer.add_char buf '-';
          last_char_is_dash := false );
        Buffer.add_char buf c
      | _ -> last_char_is_dash := true)
    s;
  Buffer.contents buf
```

CraigFe commented:
I like this idea, but it has the snag of `slugify` not being injective: what will happen if multiple test names get mapped to the same directory? (such as if someone were to use emoji for all of their test names).

samoht commented:
good point -- maybe just add a unique id/counter at the beginning of the filename.

CraigFe commented:
Sounds good to me. I suggest adding the UID at the end of the filename, so that it doesn't mess with tab-completion.

emillon commented:
This may be a naive question, but why do tests write to the disk?

CraigFe commented:
The log output of every test is written to disk separately, I assume as a debugging mechanism

Now that you mention it, this should potentially be an opt-in feature? I just checked my disk, and each Irmin `runtest` writes 8.5MB of logs, even if everything is passing.

samoht commented:
You need a way to keep track of logs to show them back to the user in case some of them fail -- to inspect CI issues, etc.
